rpc/src/endpoint/validators.rs

```diff
@@ -29,12 +66,28 @@ impl crate::SimpleRequest for Request {}
 
 /// Validator responses
 #[derive(Clone, Debug, Deserialize, Serialize)]
+#[non_exhaustive]
```

Comment:
Can you explain why we need/what does this attribute brings?

Reply:
I'm in two minds about this right now.

On the one hand, I like the fact that it reduces the number of ways in which to construct a struct (this reduces the number of ways in which we can break downstream projects' code). If we restrict construction to the constructor, it also then, for certain structs, allows for parameter validation and additional guarantees that the resulting struct is most likely "correct".

On the other hand, we have some pretty massive structs where this doesn't work so well, where they have 9+ public fields. So the constructor-based approach becomes a little unwieldy. Usually I'd then use some builder pattern to construct such structs, but this will result in a significant amount more code (where I'm trying to avoid bloat, since if there's less code it's quicker to change).

I suppose, erring on the side of rapid development and less code (which is probably a good goal for the short- to medium-term), it's probably a better idea to go with the approach of direct struct construction?

Reply:
a)

> On the one hand, I like the fact that it reduces the number of ways in which to construct a struct (this reduces the number of ways in which we can break downstream projects' code). If we restrict construction to the constructor, it also then, for certain structs, allows for parameter validation and additional guarantees that the resulting struct is most likely "correct".

Agreed, I think this is the right approach for structs that are exposed publicly and which we expect our users to instantiate themselves manually. But, even if possible, it's not really the case here as we expect users to use the higher-level methods rather than use `perform` and construct the requests directly.

b)

> On the other hand, we have some pretty massive structs where this doesn't work so well, where they have 9+ public fields. So the constructor-based approach becomes a little unwieldy. Usually I'd then use some builder pattern to construct such structs, but this will result in a significant amount more code (where I'm trying to avoid bloat, since if there's less code it's quicker to change).

If we ever want to use the builder pattern everywhere, this may help: https://lib.rs/crates/derive_builder :)

c)

> I suppose, erring on the side of rapid development and less code (which is probably a good goal for the short- to medium-term), it's probably a better idea to go with the approach of direct struct construction?

Seems fine to me for now, given the reasoning I point a) above.

Reply:
`derive_builder` looks amazing. We should definitely investigate using that approach for all the internal Tendermint domain types!